### PR TITLE
API tests

### PR DESCRIPTION
**Pull request by egorras**
https://github.com/Likvido/Likvido.App/issues/10075

Extracted shared code for API tests:
* Taken most of the implementation from `Likvido.WebAppApi.Tests`
* Added option for separate design/runtime contexts - migrate/testdata/cleanup on design context, execute tests on runtime context
* Sqlite db fixture (some changes from `Likvido.Payables.Api.Tests`)
* In usage can inherit from `ConfigurableTestFixture` and define mocks (auto configured) and named http clients (e.g. auth with api key)

### CODE REVIEW

File: src/Likvido.Test.Api/Base/ConfigurableTestFixture.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+﻿using System;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Likvido.Test.Api
+{
+    public abstract class ConfigurableTestFixture<TStartup, TDesignContext, TRuntimeContext>
+        : TestFixture<TStartup, TDesignContext, TRuntimeContext>
+        where TStartup : class
+        where TDesignContext : TRuntimeContext
```
Comment (Boychenko) [inline src/Likvido.Test.Api/Base/ConfigurableTestFixture.cs]: I think this constraint is too specific 
```suggestion
        where TDesignContext : DbContext
```

File: src/Likvido.Test.Api/Base/FixtureOptions.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+using System.Collections.Generic;
+using System.Net.Http;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Likvido.Test.Api
+{
+    public class FixtureOptions<TDesignContext, TRuntimeContext>
+        where TDesignContext : TRuntimeContext
```
Comment (Boychenko) [inline src/Likvido.Test.Api/Base/FixtureOptions.cs]: same here

File: src/Likvido.Test.Api/Base/TestFixture.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                }
+
+                return _httpClientFactory;
+            }
+
+        }
+
+        public virtual HttpClient Client => HttpClientFactory.HttpClient;
+
+        public TRuntimeContext Context
```
Comment (Boychenko) [inline src/Likvido.Test.Api/Base/TestFixture.cs]: Shouldn't we with design-time context during arrange/assert phase?

File: src/Likvido.Test.Api/DbFixtures/ContextFixtureBase.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using Microsoft.EntityFrameworkCore;
+
+namespace Likvido.Test.Api
+{
+    public abstract class ContextFixtureBase<TDesignContext, TRuntimeContext> : IDisposable
+        where TDesignContext : TRuntimeContext
```
Comment (Boychenko) [inline src/Likvido.Test.Api/DbFixtures/ContextFixtureBase.cs]:
```suggestion
        where TDesignContext : DbContext
```

File: src/Likvido.Test.Api/DbFixtures/DatabaseFixture.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        public abstract ContextFixtureBase<TDesignContext, TRuntimeContext> ContextFixture { get; }
+
+        public TRuntimeContext Context => ContextFixture.RuntimeContext;
+
+        public abstract DatabaseFixtureType DbType { get; }
+
+        public TRuntimeContext CreateRuntimeContext()
+        {
+            return ContextFixture.Create().Runtime;
```
Comment (Boychenko) [inline src/Likvido.Test.Api/DbFixtures/DatabaseFixture.cs]: Doesn't look like to create design-time context here and never using it

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc.Testing;
+using Microsoft.AspNetCore.TestHost;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Likvido.Test.Api
+{
+
```
Comment (Boychenko) [inline src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]: extra line

File: src/Likvido.Test.Api/Base/TestFixture.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                .Where(x => typeof(Mock).IsAssignableFrom(x.PropertyType))
+                .Select(x => (x.PropertyType.GetGenericArguments()[0], (Mock)x.GetValue(this)!))
+                .ToList();
+
+            _mocks.ForEach(x => services.AddSingleton(x.ServiceType, x.Mock.Object));
+        }
+
+        private void ResetMocks()
+        {
+            _mocks?.ForEach(x => x.Mock.Reset());
```
Comment (Boychenko) [inline src/Likvido.Test.Api/Base/TestFixture.cs:124]: Let's allow library consumers to provide default behavior

**Review by Boychenko — CHANGES_REQUESTED** [review]

File: src/Likvido.Test.Api/Base/TestFixture.cs (hunk 3 above)
Comment (egorras) [reply src/Likvido.Test.Api/Base/TestFixture.cs]: Updated to use design

File: src/Likvido.Test.Api/Base/TestFixture.cs (hunk 7 above)
Comment (egorras) [reply src/Likvido.Test.Api/Base/TestFixture.cs:124]: Allowed to override reset. Changed calls a bit, so reset  is called on started and moved out mocks init

File: src/Likvido.Test.Api/DbFixtures/Sqlite/SqliteContextFixture.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            var options = optionsBuilder
+                .UseSqlite(_connection)
+                .ReplaceService<IMigrationsSqlGenerator, SqliteMigrationGenerator>()
+                .Options;
+            return (TDesignContext)Activator.CreateInstance(typeof(TDesignContext), options)!;
+        }
+
+        public override void ExecuteMigration(TDesignContext context)
+        {
+            context.Database.EnsureCreated();
```
Comment (egorras) [inline src/Likvido.Test.Api/DbFixtures/Sqlite/SqliteContextFixture.cs:46]: No luck with `Migrate`, does not seem to work.
Also `SqliteMigrationGenerator` does get called on `EnsureCreated`

**Comment by egorras** [comment]
Usage example https://github.com/Likvido/Likvido.Payables/pull/753

File: src/Likvido.Test.Api/DbFixtures/ContextFixtureBase.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+using Microsoft.EntityFrameworkCore;
+
+namespace Likvido.Test.Api
+{
+    public abstract class ContextFixtureBase<TDesignContext, TRuntimeContext> : IDisposable
+        where TDesignContext : DbContext
+        where TRuntimeContext : DbContext
+    {
+        private TDesignContext? _designContext;
+        private TRuntimeContext? _runtimeContext;
```
Comment (Boychenko) [inline src/Likvido.Test.Api/DbFixtures/ContextFixtureBase.cs]: it seems to me that we don't need this field

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactoryOptions.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Likvido.Test.Api
+{
+    public class HttpClientFactoryOptions
+    {
+        public Action<IServiceCollection>? ConfigureServices { get; set; }
+#pragma warning disable CA2227 // Collection properties should be read only
+        public Dictionary<string, Action<HttpClient, IConfiguration>>? ConfigureNamedHttpClients { get; set; }
```
Comment (Boychenko) [inline src/Likvido.Test.Api/HttpClientFactory/HttpClientFactoryOptions.cs]: 1. `IServiceProvider` gives much better flexibility
2. I'm suggesting to change parameters order to match `AddHttpClient` extension method
```suggestion
        public Dictionary<string, Action<IServiceProvider, HttpClient>>? ConfigureNamedHttpClients { get; set; }
```

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        public HttpClient GetNamedHttpClient(string name)
+        {
+            if (_namedHttpClients.TryGetValue(name, out var namedHttpClient))
+            {
+                return namedHttpClient;
+            }
+
+            if (_options.ConfigureNamedHttpClients != null &&
+                _options.ConfigureNamedHttpClients.TryGetValue(name, out var httpClientConfig))
```
Comment (Boychenko) [inline src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]:
```suggestion
            if (_options?.ConfigureNamedHttpClients?.TryGetValue(name, out var httpClientConfig) == true)
```

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+            {
+                var httpClient = _factory.CreateClient();
+                httpClientConfig.Invoke(httpClient, Configuration);
```
Comment (Boychenko) [inline src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]:
```suggestion
                httpClientConfig?.Invoke(httpClient, Configuration);
```

**Review by Boychenko — CHANGES_REQUESTED** [review]

File: src/Likvido.Test.Api/DbFixtures/ContextFixtureBase.cs (hunk 9 above)
Comment (egorras) [reply src/Likvido.Test.Api/DbFixtures/ContextFixtureBase.cs]: removed

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactoryOptions.cs (hunk 10 above)
Comment (egorras) [reply src/Likvido.Test.Api/HttpClientFactory/HttpClientFactoryOptions.cs]: 👍 updated

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs (hunk 11 above)
Comment (egorras) [reply src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]: This leads to `httpClientConfig` considered as not initialized 😢

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs (hunk 12 above)
Comment (egorras) [reply src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]: 👍

Comment (Boychenko) [reply src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]: Can't this this change 👀

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        public HttpClient HttpClient => _httpClient ??= _factory.CreateClient();
+
+        public HttpClient GetNamedHttpClient(string name)
+        {
+            if (_namedHttpClients.TryGetValue(name, out var namedHttpClient))
+            {
+                return namedHttpClient;
+            }
+
+            if (_options.ConfigureNamedHttpClients != null &&
```
Comment (Boychenko) [inline src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]:
```suggestion
            if (_options?.ConfigureNamedHttpClients != null &&
```

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs (hunk 11 above)
Comment (Boychenko) [reply src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]: check another suggestion then

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs (hunk 12 above)
Comment (egorras) [reply src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]: Missed commit somehow, added

File: src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs (hunk 13 above)
Comment (egorras) [reply src/Likvido.Test.Api/HttpClientFactory/HttpClientFactory.cs]: done

**Review by Boychenko — APPROVED** [review]

**Pull request merged**